### PR TITLE
Send emails automatically to applicants

### PR DESCRIPTION
**Pull request by takiarek**
https://trello.com/c/DEsb6WCV/32-mailers-send-emails-automatically-to-applicants-accepted-rejected-and-on-waitlist

This PR is here just for history and discussion. There is an organized version of this PR: https://github.com/LunarLogic/rails-girls-submissions/pull/60.

### CODE REVIEW

File: app/services/submissions_expiration_handler.rb [hunk 1]
```diff
@@ -0,0 +1,36 @@
+class SubmissionsExpirationHandler
+  def self.build
+    new(SubmissionRepository.new, SubmissionsInviter.new)
+  end
+
+  def initialize(submission_repository, submissions_inviter)
+    @submission_repository = submission_repository
+    @submissions_inviter = submissions_inviter
+  end
```
Comment (marekciupak) [inline app/services/submissions_expiration_handler.rb]: I find using `self.build` or `self.call` useful when I need to provide some input data and I do not want to play with fake repositories in tests. Example:

```ruby
def self.call(date:, measurements_repository: MeasurementsRepository.new)
  new(input_data: measurements_repository.for_day(date)).call
end

def initialize(input_data:)
  @input_data: input_data
end
```

This way I can easily inject data for my object in tests `Service.new(input_data: [...])`, without creating a fake repo. But, I can also use `Service.call(date: Date.today)` in the other services.

In your case, I don't see value for creating separate `self.build` method. I think you can just write it this way:
```ruby
def initialize(submission_repository: SubmissionRepository.new, submissions_inviter: SubmissionsInviter.new)
  @submission_repository = submission_repository
  @submissions_inviter = submissions_inviter
end
```

File: app/services/submissions_expiration_handler.rb [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+
+  def submissions_to_be_expired
+    Submission.select do |submission|
+      submission.has_expired?
+    end
+  end
+
+  def invite_submission(waitlist_submission)
+    @submissions_inviter.invite(waitlist_submission)
```
Comment (marekciupak) [inline app/services/submissions_expiration_handler.rb]: IMO, it is better to use `attr_reader` instead of `@`. You can easier change `attr_reader` into a custom method if needed in the future.

```ruby
private
attr_reader: :submissions_inviter
```

File: config/routes.rb [hunk 3]
```diff
@@ -5,11 +5,14 @@
   get "/submissions/rejected", to: "submissions#rejected"
   get "/submissions/to_rate", to: "submissions#to_rate"
   get "/submissions/results", to: "submissions#results"
+  get "/submissions/confirm", to: "submissions#confirm"
```
Comment (marekciupak) [inline config/routes.rb]: I would use `put` instead of `get`

File: config/routes.rb [hunk 4]
```diff
@@ -5,11 +5,14 @@
...
 
   get "/csv/download_accepted", to: "csv#download_accepted"
   get "/csv/download_waitlist", to: "csv#download_waitlist"
   get "/csv/download_unaccepted", to: "csv#download_unaccepted"
 
+  get "/send_invitation_emails", to: "mailings#send_invitation_emails"
```
Comment (marekciupak) [inline config/routes.rb]: I would use `post` instead of `get`

File: app/models/submission.rb [hunk 5]
```diff
@@ -32,5 +34,14 @@ def reject
...
 
+  def generate_confirmation_token!
+    self.confirmation_token = Devise.friendly_token
+    self.confirmation_token_created_at = Time.now
+    save
+  end
+
+  def has_expired?
+    awaiting? && confirmation_token_created_at < 1.week.ago
+  end
```
Comment (marekciupak) [inline app/models/submission.rb]: I think this method with that name is misleading. If some submission is already marked as expired, then `submission.has_expired?` will return `false`.

You could change the name to `to_be_marked_as_expired`. Personally, I would move the logic into the service which is using that method.

File: spec/features/inviting_accepted_submissions_feature_spec.rb [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+      :submission, 2,
+      :with_rates,
+      confirmation_status: 'not_avaible',
+      rates_num: setting.required_rates_num,
+      rates_val: 1)
+
+    login_as(user, scope: :user)
+    visit submissions_results_path
+    click_link('Send')
+    click_link('Send')
```
Comment (marekciupak) [inline spec/features/inviting_accepted_submissions_feature_spec.rb]: Does each user have a separate button?

In such cases, I tend to set CSS class with user id for a button, so I can easily identify the correct one. In your case, it may be not important, b/c you want to simply click all of them, but I wanted to share it.

File: spec/features/inviting_accepted_submissions_feature_spec.rb [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+      confirmation_token_created_at: 1.week.ago - 1,
+      rates_num: setting.required_rates_num,
+      rates_val: 1)
+
+    visit submissions_confirm_path(confirmation_token: invited_submission.confirmation_token)
+    expect(invited_submission.reload).not_to be_confirmed
+    expect(page).to have_text('expired')
+  end
+
+  it 'show correct info when invitation is already confirmed but token expired' do
```
Comment (marekciupak) [inline spec/features/inviting_accepted_submissions_feature_spec.rb]: good one :)

File: spec/services/submissions_inviter_spec.rb [hunk 8]
```diff
@@ -0,0 +1,17 @@
+require 'rails_helper'
+
+describe SubmissionsInviter do
+  describe '#invite' do
+    let(:submission) { instance_double Submission }
+    let(:message_delivery) { instance_double ActionMailer::MessageDelivery }
```
Comment (marekciupak) [inline spec/services/submissions_inviter_spec.rb]: I wouldn't mock rails code, I would just check `ActionMailer::Base.deliveries.count` after sending the email.

File: app/controllers/mailings_controller.rb [hunk 9]
```diff
@@ -0,0 +1,11 @@
+class MailingsController < ApplicationController
+  def send_invitation_emails
+    submissions_to_invite = SubmissionRepository.new.to_invite
+
+    submissions_to_invite.each do |submission|
+      SubmissionsInviter.new.invite(submission)
+    end
+
+    redirect_to :back, notice: "You have sent the emails."
```
Comment (czorek) [inline app/controllers/mailings_controller.rb]: I would specify the message to say `You have sent the invitation emails.`

File: app/models/submission.rb [hunk 10]
```diff
@@ -32,5 +34,14 @@ def reject
...
   end
-end
 
+  def generate_confirmation_token!
+    self.confirmation_token = Devise.friendly_token
+    self.confirmation_token_created_at = Time.now
+    save
+  end
+
+  def has_expired?
```
Comment (czorek) [inline app/models/submission.rb]: I would rename this method, because it doesn't tell me what is expired - invitation or submission?

File: app/repositories/submission_repository.rb [hunk 11]
```diff
@@ -31,6 +31,10 @@ def waitlist
     rated_scope.offset(Setting.get.available_spots).to_a
   end
 
+  def to_invite
+    rated_scope.where.not(confirmation_status: 3).limit(Setting.get.available_spots).where(confirmation_status: 0).to_a
```
Comment (czorek) [inline app/repositories/submission_repository.rb]: Or constants

File: app/controllers/submissions_controller.rb [hunk 12]
```diff
@@ -1,6 +1,17 @@
 class SubmissionsController < ApplicationController
-  skip_before_action :authenticate_user!, only: [:new, :create, :thank_you]
-  layout 'dashboard', only: [:valid, :rated, :to_rate, :rejected, :results]
+  skip_before_action :authenticate_user!, only: [:confirm, :new, :create, :thank_you]
+  layout 'dashboard', only: [:valid, :all, :rated, :to_rate, :rejected, :results]
+
+  def confirm
+    token = params[:confirmation_token]
+    submission = Submission.find_by(confirmation_token: token)
+    if submission.has_expired?
```
Comment (ArturT) [inline app/controllers/submissions_controller.rb]: or you could raise exception `Submission.find_by!`

using `if submission.try(&:has_expired?)` will trigger `else` and `submission.confirmed!` will be called on `nil`

File: app/services/submissions_expiration_handler.rb (hunk 2 above)
Comment (czorek) [inline app/services/submissions_expiration_handler.rb]: Same with `@submission_repository`

File: app/services/submissions_expiration_handler.rb [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+  end
+
+  private
+
+  def waitlist_submissions
+    @submission_repository.to_invite
+  end
+
+  def submissions_to_be_expired
+    Submission.select do |submission|
```
Comment (czorek) [inline app/services/submissions_expiration_handler.rb]: Shouldn't this also be moved to the submission repository?

File: app/models/submission.rb (hunk 5 above)
Comment (ArturT) [inline app/models/submission.rb]: or maybe you could call it `can_expire?`

File: spec/features/inviting_accepted_submissions_feature_spec.rb (hunk 6 above)
Comment (czorek) [inline spec/features/inviting_accepted_submissions_feature_spec.rb]: I don't get why are there 2 clicks - we have one button for all invitations, right? That''s how it looks in the view

File: app/controllers/submissions_controller.rb [hunk 14]
```diff
@@ -1,6 +1,17 @@
...
+      render text: "Time for confirmation expired!"
+    else
+      submission.confirmed!
```
Comment (takiarek) [inline app/controllers/submissions_controller.rb]: It's method from rails enum for confirmation_status. Do you recommend just wrapping this method

`def confirm!
    self.confirmed!
  end`

or maybe setting confirmation_status explicitly:

`def confirm!
    self.confirmation_status = 'confirmed'
    save
  end`

Comment (takiarek) [inline app/controllers/submissions_controller.rb]: Or maybe leaving it as it is since it's from enum?

File: spec/models/submission_spec.rb [hunk 15]
```diff
@@ -18,4 +18,44 @@
...
+        :submission,
+        confirmation_status: 'expired',
+        confirmation_token_created_at: 1.week.ago - 1)
+      not_invited_submission = FactoryGirl.build(
+        :submission,
+        confirmation_status: 'not_avaible',
+        confirmation_token_created_at: 1.week.ago - 1)
+      submissions = [confirmed_submission, expired_submission, not_invited_submission]
+
+      expect(confirmed_submission).not_to be_has_expired
```
Comment (czorek) [inline spec/models/submission_spec.rb]: This reads bad, I would probably write `expect(confirmed_submission.has_expired?).to be_false`

File: spec/repositories/submission_repository_spec.rb [hunk 16]
```diff
@@ -109,4 +109,36 @@
       end
     end
   end
+
+  describe '#to_invite' do
```
Comment (czorek) [inline spec/repositories/submission_repository_spec.rb]: Do we really have to create 5 full-fledged objects here?

File: spec/features/inviting_accepted_submissions_feature_spec.rb (hunk 6 above)
Comment (takiarek) [inline spec/features/inviting_accepted_submissions_feature_spec.rb]: There is one button for all submissions.

File: app/controllers/submissions_controller.rb (hunk 12 above)
Comment (ArturT) [inline app/controllers/submissions_controller.rb]:
```ruby
2.3.0 :009 >   submission = nil
 => nil
2.3.0 :010 > if submission.try(:has_expired?)
2.3.0 :011?>   puts 'A'
2.3.0 :012?>   else
2.3.0 :013 >     puts 'B'
2.3.0 :014?>   end
B
=> nil
```

File: spec/features/inviting_accepted_submissions_feature_spec.rb (hunk 6 above)
Comment (takiarek) [inline spec/features/inviting_accepted_submissions_feature_spec.rb]: @czorek It's to ensure that everyone gets only one email even if the button is clicked more than once.
I can see where the confusion comes from, I will improve the code to make it more clear about what it does.

File: spec/features/inviting_accepted_submissions_feature_spec.rb (hunk 7 above)
Comment (takiarek) [inline spec/features/inviting_accepted_submissions_feature_spec.rb]: Thanks!

File: spec/features/inviting_accepted_submissions_feature_spec.rb (hunk 6 above)
Comment (marekciupak) [inline spec/features/inviting_accepted_submissions_feature_spec.rb]: Ah, ok. You could check such more complicated case in unit test.

File: app/controllers/submissions_controller.rb (hunk 12 above)
Comment (marekciupak) [inline app/controllers/submissions_controller.rb]: Yes, you're right, Artur.

File: app/models/submission.rb (hunk 5 above)
Comment (takiarek) [inline app/models/submission.rb]: @marekciupak @ArturT @czorek 
past_confirmation_due_date?
past_confirmation_period?

The 'confirmation' term makes some trouble with wording, maybe it would be better to call the status "invitation_status" instead, although I wouldn't make such a change until we consider @marekciupak's suggestion about dealing with this without additional status whatsoever.

to_be_marked_as_expired? - this method is used for more than that

File: spec/repositories/submission_repository_spec.rb (hunk 16 above)
Comment (takiarek) [inline spec/repositories/submission_repository_spec.rb]: My aim here was to ensure that this method returns correct submissions in every possible scenario so I don't need to test edge cases every time another class uses this method. Is there a better way to achieve that? Is there a better approach to testing things like this? 
Maybe bad code design is the source of this crammed test?

Comment (marekciupak) [inline spec/repositories/submission_repository_spec.rb]: Assuming that you decide to calculate status dynamically and you have `confirmation_token_generated_at` and `confirmed_at` instead of `status` attribute:

these 3 objects would be enough for me:
* `to_invite` with high `rate_val` and `confirmation_token_generated_at: nil`
* `invited` with high `rate_val` and `confirmation_token_generated_at: 6.days.ago`
* `not_accepted` with too low `rate_val`

`SubmissionRepository#to_invite` should return only `to_invite` one.

File: app/repositories/submission_repository.rb (hunk 11 above)
Comment (takiarek) [inline app/repositories/submission_repository.rb]: It won't work with enums. I'm gonna do this: `.where('confirmation_status <> ?', Submission.confirmation_statuses[:expired])`

File: spec/services/submissions_inviter_spec.rb (hunk 8 above)
Comment (takiarek) [inline spec/services/submissions_inviter_spec.rb]: What if I want to ensure that this service uses the mailer I want it to use not just that it sends an email?

File: app/controllers/submissions_controller.rb [hunk 17]
```diff
@@ -1,6 +1,19 @@
...
+
+  def confirm
+    token = params.require(:confirmation_token)
+    submission = Submission.find_by!(confirmation_token: token)
+    if submission.past_confirmation_due_date?
+      render text: "Time for confirmation expired!"
+    else
+      submission.confirmed!
+      render text: "You confirmed your invitation!"
+    end
```
Comment (lesniakania) [inline app/controllers/submissions_controller.rb]: Not necessarily, but personally I'd probably move those action to service. But as there are not so much of them, I don't have strong opinion about this.

File: app/mailers/application_mailer.rb [hunk 18]
```diff
@@ -0,0 +1,4 @@
+class ApplicationMailer < ActionMailer::Base
+  default from: "test@example.com"
```
Comment (lesniakania) [inline app/mailers/application_mailer.rb]: it would be good to have this email in config. I see we don't have a separate config, but we use `secrets.yml` for that, so you can add it there.

File: app/repositories/submission_repository.rb [hunk 19]
```diff
@@ -31,6 +31,14 @@ def waitlist
     rated_scope.offset(Setting.get.available_spots).to_a
   end
 
+  def to_invite
+    rated_scope
+      .where('confirmation_status <> ?', Submission.confirmation_statuses[:expired])
+      .limit(Setting.get.available_spots)
+      .where('confirmation_status = ?', Submission.confirmation_statuses[:not_available])
+      .to_a
```
Comment (lesniakania) [inline app/repositories/submission_repository.rb]: not sure if this `to_a` is needed here, without this it could be more flexible (you could add more conditions or something if you want)

File: app/services/submissions_expiration_handler.rb [hunk 20]
```diff
@@ -0,0 +1,29 @@
...
+    submissions_to_be_expired.each do |s|
+      s.expired!
+    end
+  end
+
+  def submissions_to_be_expired
+    Submission.select do |submission|
+      submission.past_confirmation_due_date?
+    end
+  end
```
Comment (lesniakania) [inline app/services/submissions_expiration_handler.rb]: personally, I would probably move expiring to the separate service, to be consistent with `SubmissionsInviter` and just use `submisions_expirer.call` and `submissions_inviter.call` in this service `call` method.

But for sure I would move retrieving submissions to be expired to the repository and use database for that.

File: app/services/submissions_inviter.rb [hunk 21]
```diff
@@ -0,0 +1,20 @@
+class SubmissionsInviter
+  def initialize(submission_repository: SubmissionRepository.new)
+    @submission_repository = submission_repository
+  end
```
Comment (lesniakania) [inline app/services/submissions_inviter.rb]: I like that you use dependency injection here and in `SubmissionsExpirationHandler` 👍

File: config/routes.rb (hunk 3 above)
Comment (lesniakania) [inline config/routes.rb]: Initially, I wanted to agree, but then I thought that it's needed for the link in the email, so it's like email confirmation or something. I think `get` is ok.

File: spec/tasks/handle_submissions_expiration_spec.rb [hunk 22]
```diff
@@ -0,0 +1,31 @@
...
+
+    Rake::Task.define_task(:environment)
+  end
+
+  it 'calls correct service' do
+    expect(SubmissionsExpirationHandler).to receive(:new).and_return(submissions_expiration_handler)
+    expect(submissions_expiration_handler).to receive(:call)
+    subject.invoke
+  end
+end
```
Comment (lesniakania) [inline spec/tasks/handle_submissions_expiration_spec.rb]: I think it's first time in my life when I see spec for rake task ; d Nice ; )

File: app/services/submissions_expiration_handler.rb [hunk 23]
```diff
@@ -0,0 +1,29 @@
...
+    expire_submissions
+    invite_submissions
+  end
+
+  private
+  attr_reader :submissions_inviter
+
+  def expire_submissions
+    submissions_to_be_expired.each do |s|
+      s.expired!
```
Comment (lesniakania) [inline app/services/submissions_expiration_handler.rb]: I think it would be good to get an email when your submission expired

File: app/views/layouts/mailer.text.erb [hunk 24]
```diff
@@ -0,0 +1 @@
+<%= yield %>
```
Comment (hanka) [inline app/views/layouts/mailer.text.erb]: since you have layout for emails in text format maybe it's also worth adding emails content as well ;)

File: db/migrate/20161228084444_add_invitation_fields_to_submissions.rb [hunk 25]
```diff
@@ -0,0 +1,6 @@
+class AddInvitationFieldsToSubmissions < ActiveRecord::Migration
+  def change
+    add_column :submissions, :invitation_token, :string
```
Comment (hanka) [inline db/migrate/20161228084444_add_invitation_fields_to_submissions.rb]: there should be index added for `invitation_token`

File: spec/models/submission_spec.rb [hunk 26]
```diff
@@ -18,4 +18,50 @@
...
+  describe '#invitation_expired?' do
+    let(:expired_submission) do
+      FactoryGirl.build(
+        :submission,
+        invitation_token: 'xxx',
+        invitation_token_created_at: 1.week.ago - 1,
+        invitation_confirmed: false)
+    end
+    let(:confirmed_submission) do
+      FactoryGirl.build(
```
Comment (hanka) [inline spec/models/submission_spec.rb]: picking on, but be consistent in formatting ;)

File: spec/features/inviting_new_submissions_in_place_of_expired_ones_feature_spec.rb [hunk 27]
```diff
@@ -0,0 +1,40 @@
+# Testing rake tasks method found at:
+# http://carlosplusplus.github.io/blog/2014/02/01/testing-rake-tasks-with-rspec/
+# https://robots.thoughtbot.com/test-rake-tasks-like-a-boss
+
+require 'rails_helper'
+require 'rake'
+
+describe 'inviting new submissions in place of expired ones' do
```
Comment (czorek) [inline spec/features/inviting_new_submissions_in_place_of_expired_ones_feature_spec.rb]: So you do all this stuff, 30+ lines of code, with a setup that is tightly coupled to rails and code that is not yours, just to check if it calls the correct service? And you have a spec for that service as well? This seems like a huge test overkill to me. I know how you like to test things thoroughly, but just because you can test something, doesn't mean you should ;)

File: spec/features/reminding_about_expiring_invitations_feature_spec.rb [hunk 28]
```diff
@@ -0,0 +1,38 @@
+require 'rails_helper'
+require 'rake'
+
+describe 'reminding about expiring invitations' do
```
Comment (czorek) [inline spec/features/reminding_about_expiring_invitations_feature_spec.rb]: Same as above

File: spec/features/inviting_new_submissions_in_place_of_expired_ones_feature_spec.rb (hunk 27 above)
Comment (takiarek) [inline spec/features/inviting_new_submissions_in_place_of_expired_ones_feature_spec.rb]: I would like to live in the world where running your test suite is all you need to catch regressions. Without those tests, it's easy to forget about updating rake tasks when service's interface changes. I see two options for improvement here:
1. The config can be extracted to shared example and will work for each rake task spec.
2. I can remove those tests completely because when changing interface you should probably grep the repo anyway to double check.
What do you think?

Comment (takiarek) [inline spec/features/inviting_new_submissions_in_place_of_expired_ones_feature_spec.rb]: Another argument against keeping those tests is that we're still not 100% covered because after changing name o the rake task, you can forget to update schedule.rb which is not tested. Unless we test schedule.rb... :smiling_imp:

Comment (czorek) [inline spec/features/inviting_new_submissions_in_place_of_expired_ones_feature_spec.rb]: Well, the service's interface is `#call`, I don't think that this would be changed a lot or ever. And if somebody changes the interface or object name and doesn't grep for all the places he should change it at - he deserves all the errors he gets. Besides, I don't think you can ever test an app 100%, there will always be scenarios where something can fail and tests won't catch that. You should be able to live with that and say "good enough" sometimes :)

File: db/migrate/20161228084444_add_invitation_fields_to_submissions.rb (hunk 25 above)
Comment (lesniakania) [inline db/migrate/20161228084444_add_invitation_fields_to_submissions.rb]: I agree with Hania
